api/v1alpha7/types.go

```diff
@@ -163,6 +163,39 @@ type RootVolume struct {
 	AvailabilityZone string `json:"availabilityZone,omitempty"`
 }
 
+type AdditionalBlockDevice struct {
+	// BlockDeviceType is the type of block device to create.
+	// This can be either "volume" or "local".
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Enum=volume;local
+	BlockDeviceType string `json:"blockDeviceType"`
+
+	// Here I'm hesitant to add GuestFormat. Not to let users choose the format of the block device
```

Comment:
question for reviewers.

Same question for `BlockDeviceType`, do we want to support things like Snapshots, Images, etc?
In other words, do we want to support any block device type that nova/cinder can do?



Task | openstack server create parameter(s) | Information
-- | -- | --
Boot an instance from an image and attach a non-bootable volume. | --block-device | Boot_instance_from_image_and_attach_non-bootable_volume
Create a volume from an image and boot an instance from that volume. | --boot-from-volume and --image; --block-device | Create_volume_from_image_and_boot_instance
Boot from an existing source image, volume, or snapshot. | --volume or --snapshot; --block-device | Create_volume_from_image_and_boot_instance
Attach a swap disk to an instance. | --swap | Attach_swap_or_ephemeral_disk_to_an_instance
Attach an ephemeral disk to an instance. | --ephemeral | Attach_swap_or_ephemeral_disk_to_an_instance


Source: https://docs.openstack.org/nova/latest/user/launch-instance-from-volume.html

Reply:
cc @mdbooth @mkjpryor

Reply:
It depends how far we want to go with the `BlockDeviceType` abstraction vs exposing `source_type`, `destination_type`, and `guest_format` directly.

Looking at [what is currently possible](https://docs.openstack.org/nova/2023.2/user/block-device-mapping.html#valid-source-destination-combinations), we could have the following device types:
- `volume`:
  - source: volume
  - destination: volume
  - uuid: required
- `snapshot`:
  - source_type: snapshot
  - destination_type: volume
  - uuid: required
- `image`:
  - source_type: image
  - destination_type: volume
  - uuid: required
- `local`:
  - source_type: blank
  - destination_type: local
  - guest_format: whatever the user wants
  - size: required
- `swap`:
  - source_type: blank
  - destination_type: local
  - guest_format: swap
  - size: required

Swap is just a local block device, with `guest_format=swap`, so we may not want to create an additional type.

Reply:
And of course, adding support for the `snapshot` and `image` types belongs to separate PRs, if that is something we want.

Reply:
I asked @mdbooth if we could just replace the root volume stuff with `source_type: image` and he seemed quite against it. I think I understand why, but I feel that we probably won't support the image source type for additional devices (i.e. non-boot)? Not sure what the use case for that would be.

Reply:
`swap` might be useful to support in the future, now that Kubernetes has a bit more support for running with swap on.

Reply:
It's still not encouraged though, so I'm not sure we should support it... 🤷‍♂️

Reply:
> replace the root volume stuff

Do you mean, removing `rootVolume` and make `additionalBlockDevice` supporting `source_type: image` like you said?
I wouldn't be against but this seems like yet another API change (although conversion would be possible).

Also in that case we would maybe remove the "Additional" from the field name, and have `BlockDevices` ?

> swap

I agree it's useful for users who know what they're doing. I wonder if we can support that later though or if it needs to be part of that API for now.

Reply:
I don't want to proxy the OpenStack API. Every time we've done that so far it has created an almighty mess. We need to understand our own use cases and present them to the user in a way which makes sense in our context. Specifically I would not be in favour of directly exposing `source_type`, `destination_type`, and `guest_format`.

I agree that we don't have a current use case for swap or snapshot, and I don't see any reason to support them until we do.

My primary concern here is actually the supportability of the 'multiple ephemeral disks' feature in Nova, which this API implies. When I maintained this code it was my impression that it was not well tested or widely used. Most users have a single ephemeral disk. Being possibly the only user of this feature makes me nervous.

I could be persuaded that it's ok if we test it ourselves on both current OpenStack and the oldest OpenStack we want to support. If it currently works in our window, I think we can trust OpenStack to fix regressions if they arise.

Reply:
> we don't have a current use case for swap or snapshot, and I don't see any reason to support them until we do.

:+1:

> the supportability of the 'multiple ephemeral disks' feature in Nova

I think this was asked by @mkjpryor who has a valid use case.
I would say that we add it and maybe document that this is tested as best effort in our CI (for the `etcd` use case ultimately) but that there are some risks due to OpenStack upstream testing coverage.

Reply:
Just had a chat with @mdbooth about `swap`, he seems against it.
If we want users to add swap, we'll have to add a check for `swap` string in `guestFormat`. Is that what we want?

cc @mandre

Reply:
A chat with @mdbooth today:
* We don't want to prevent users to deploy `swap`, but we don't want to add explicit support for it.
* Nova formatted ephemeral disks have always been problematic so we want to warn our users that using `guestFormat` is at their own risk for both swap or any filesystem.